In [1]:
#Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set path

path = r'C:\Users\Josh Wattay\OneDrive\Documents\Desktop\Flight Delay Analysis - Raw Data\2020'

In [3]:
# Import Datasets

df = pd.read_csv(os.path.join(path, 'flights2020.csv'), index_col = False)

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_9136\81471578.py:3: DtypeWarning: Columns (11,52) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(os.path.join(path, 'flights2020.csv'), index_col = False)


In [5]:
#Shape Check
df.shape

(5022397, 70)

In [6]:
# Set the display option to show all 70 columns
pd.set_option('display.max_columns', 70)

In [7]:
#Inspect Dataframe and begin Cleaning by ensuring columns are named appropriately
df.head()

Unnamed: 0  Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0           0  2020        1      1          17          5  2020-01-17   
1           1  2020        1      1          18          6  2020-01-18   
2           2  2020        1      1          19          7  2020-01-19   
3           3  2020        1      1          20          1  2020-01-20   
4           4  2020        1      1          21          2  2020-01-21   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        B6                     20409   
1                        B6                     20409   
2                        B6                     20409   
3                        B6                     20409   
4                        B6                     20409   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          B6                              942   
1                          B6                              942   
2                          B6                              942   
3                          B6                              942   
4                          B6                              942   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 B6   
1                                               NaN                 B6   
2                                               NaN                 B6   
3                                               NaN                 B6   
4                                               NaN                 B6   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20409                          B6      N661JB   
1                     20409                          B6      N584JB   
2                     20409                          B6      N662JB   
3                     20409                          B6      N784JB   
4                     20409                          B6      N784JB   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                              942            14027             1402702   
1                              942            14027             1402702   
2                              942            14027             1402702   
3                              942            14027             1402702   
4                              942            14027             1402702   

   OriginCityMarketID Origin                  OriginCityName OriginState  \
0               34027    PBI  West Palm Beach/Palm Beach, FL          FL   
1               34027    PBI  West Palm Beach/Palm Beach, FL          FL   
2               34027    PBI  West Palm Beach/Palm Beach, FL          FL   
3               34027    PBI  West Palm Beach/Palm Beach, FL          FL   
4               34027    PBI  West Palm Beach/Palm Beach, FL          FL   

   OriginStateFips OriginStateName  DestAirportID  DestAirportSeqID  \
0               12         Florida          10529           1052906   
1               12         Florida          10529           1052906   
2               12         Florida          10529           1052906   
3               12         Florida          10529           1052906   
4               12         Florida          10529           1052906   

   DestCityMarketID Dest  DestCityName DestState  DestStateFips DestStateName  \
0             30529  BDL  Hartford, CT        CT              9   Connecticut   
1             30529  BDL  Hartford, CT        CT              9   Connecticut   
2             30529  BDL  Hartford, CT        CT              9   Connecticut   
3             30529  BDL  Hartford, CT        CT              9   Connecticut   
4             30529  BDL  Hartford, CT        CT              9   Connecticut   

   CRSDepTime  DepTime  DepDelay  DepDelayMinutes  DepDel15  \
0        1940   1926.0     -14.0              0.0       0.0  

The 'Unnamed: 0' column will be removed as it is an additional index and redundant. 

In [8]:
# Drop the 'Unnamed: 0' column
df = df.drop(columns=['Unnamed: 0'])

In [9]:
df.shape

(5022397, 69)

In [10]:
df.head()

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0  2020        1      1          17          5  2020-01-17   
1  2020        1      1          18          6  2020-01-18   
2  2020        1      1          19          7  2020-01-19   
3  2020        1      1          20          1  2020-01-20   
4  2020        1      1          21          2  2020-01-21   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        B6                     20409   
1                        B6                     20409   
2                        B6                     20409   
3                        B6                     20409   
4                        B6                     20409   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          B6                              942   
1                          B6                              942   
2                          B6                              942   
3                          B6                              942   
4                          B6                              942   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 B6   
1                                               NaN                 B6   
2                                               NaN                 B6   
3                                               NaN                 B6   
4                                               NaN                 B6   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     20409                          B6      N661JB   
1                     20409                          B6      N584JB   
2                     20409                          B6      N662JB   
3                     20409                          B6      N784JB   
4                     20409                          B6      N784JB   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                              942            14027             1402702   
1                              942            14027             1402702   
2                              942            14027             1402702   
3                              942            14027             1402702   
4                              942            14027             1402702   

   OriginCityMarketID Origin                  OriginCityName OriginState  \
0               34027    PBI  West Palm Beach/Palm Beach, FL          FL   
1               34027    PBI  West Palm Beach/Palm Beach, FL          FL   
2               34027    PBI  West Palm Beach/Palm Beach, FL          FL   
3               34027    PBI  West Palm Beach/Palm Beach, FL          FL   
4               34027    PBI  West Palm Beach/Palm Beach, FL          FL   

   OriginStateFips OriginStateName  DestAirportID  DestAirportSeqID  \
0               12         Florida          10529           1052906   
1               12         Florida          10529           1052906   
2               12         Florida          10529           1052906   
3               12         Florida          10529           1052906   
4               12         Florida          10529           1052906   

   DestCityMarketID Dest  DestCityName DestState  DestStateFips DestStateName  \
0             30529  BDL  Hartford, CT        CT              9   Connecticut   
1             30529  BDL  Hartford, CT        CT              9   Connecticut   
2             30529  BDL  Hartford, CT        CT              9   Connecticut   
3             30529  BDL  Hartford, CT        CT              9   Connecticut   
4             30529  BDL  Hartford, CT        CT              9   Connecticut   

   CRSDepTime  DepTime  DepDelay  DepDelayMinutes  DepDel15  \
0        1940   1926.0     -14.0              0.0       0.0   
1        1940   1926.0     -14.0              0.0       0.0   
2      

Rename syntax for various columns for consistency and accuracy.

In [11]:
# Define the dictionary for renaming columns
rename_dict = {
    'DayofMonth': 'Day_of_Month',
    'DayOfWeek': 'Day_of_Week',
    'FlightDate': 'Flight_Date',
    'Operating_Airline': 'Operating Airline',
    'OriginAirportID': 'Origin_Airport_ID',
    'OriginAirportSeqID': 'Origin_Airport_Seq_ID',
    'OriginCityMarketID': 'Origin_City_Market_ID',
    'OriginCityName': 'Origin_City_Name',
    'OriginState': 'Origin_State',
    'OriginStateFips': 'Origin_State_Fips',
    'OriginStateName': 'Origin_State_Name',
    'DestAirportID': 'Dest_Airport_ID',
    'DestAirportSeqID': 'Dest_Airport_Seq_ID',
    'DestCityMarketID': 'Dest_City_Market_ID',
    'DestCityName': 'Dest_City_Name',
    'DestState': 'Dest_State',
    'DestStateFips': 'Dest_State_Fips',
    'DestStateName': 'Dest_State_Name',
    'CRSDepTime': 'CRS_Dep_Time',
    'DepTime': 'Dep_Time',
    'DepDelay': 'Dep_Delay',
    'DepDelayMinutes': 'Dep_Delay_Minutes',
    'DepDel15': 'Dep_Del_15',
    'DepartureDelayGroups': 'Departure_Delay_Groups',
    'DepTimeBlk': 'Dep_Time_Blk',
    'TaxiOut': 'Taxi_Out',
    'WheelsOff': 'Wheels_Off',
    'WheelsOn': 'Wheels_On',
    'TaxiIn': 'Taxi_In',
    'CRSArrTime': 'CRS_Arr_Time',
    'ArrTime': 'Arr_Time',
    'ArrDelay': 'Arr_Delay',
    'ArrDelayMinutes': 'Arr_Delay_Minutes',
    'ArrDel15': 'Arr_Del_15',
    'ArrivalDelayGroups': 'Arrival_Delay_Groups',
    'ArrTimeBlk': 'Arr_Time_Blk',
    'CancellationCode': 'Cancellation_Code',
    'CRSElapsedTime': 'CRS_Elapsed_Time',
    'ActualElapsedTime': 'Actual_Elapsed_Time',
    'AirTime': 'Air_Time',
    'DistanceGroup': 'Distance_Group',
    'CarrierDelay': 'Carrier_Delay',
    'WeatherDelay': 'Weather_Delay',
    'NASDelay': 'NAS_Delay',
    'SecurityDelay': 'Security_Delay',
    'LateAircraftDelay': 'Late_Aircraft_Delay',
    'FirstDepTime': 'First_Dep_Time',
    'TotalAddGTime': 'Total_Add_G_Time',
    'LongestAddGTime': 'Longest_Add_G_Time',
    'DivDistance': 'Div_Distance'
}

# Rename the columns
df = df.rename(columns=rename_dict)

In [12]:
df.shape

(5022397, 69)

In [13]:
# Drop the additional irrelevant columns - please refer to the 'Cleaning 2019 Flight Data' script for more details.

df = df.drop(columns=['First_Dep_Time', 'Total_Add_G_Time', 'Longest_Add_G_Time', 'Div_Distance'])

Fortunately, this raw data comes with a Duplicate flag column.
I will check the counts for this column to check for any values of 'Y' or Yes.
To be certain, I will also run a duplicate check with the .duplicated command.

In [14]:
df['Duplicate'].value_counts()

Duplicate
N    5022397
Name: count, dtype: int64

In [15]:
# So far so good, now to check for duplicates in the entire dataframe by creating a subset of duplicated rows

# Create a DataFrame containing only the duplicate rows
dups = df[df.duplicated(keep=False)]

# Display the 'dups' DataFrame
print(dups)

Empty DataFrame
Columns: [Year, Quarter, Month, Day_of_Month, Day_of_Week, Flight_Date, Marketing_Airline_Network, DOT_ID_Marketing_Airline, IATA_Code_Marketing_Airline, Flight_Number_Marketing_Airline, IATA_Code_Originally_Scheduled_Code_Share_Airline, Operating_Airline , DOT_ID_Operating_Airline, IATA_Code_Operating_Airline, Tail_Number, Flight_Number_Operating_Airline, Origin_Airport_ID, Origin_Airport_Seq_ID, Origin_City_Market_ID, Origin, Origin_City_Name, Origin_State, Origin_State_Fips, Origin_State_Name, Dest_Airport_ID, Dest_Airport_Seq_ID, Dest_City_Market_ID, Dest, Dest_City_Name, Dest_State, Dest_State_Fips, Dest_State_Name, CRS_Dep_Time, Dep_Time, Dep_Delay, Dep_Delay_Minutes, Dep_Del_15, Departure_Delay_Groups, Dep_Time_Blk, Taxi_Out, Wheels_Off, Wheels_On, Taxi_In, CRS_Arr_Time, Arr_Time, Arr_Delay, Arr_Delay_Minutes, Arr_Del_15, Arrival_Delay_Groups, Arr_Time_Blk, Cancelled, Cancellation_Code, Diverted, CRS_Elapsed_Time, Actual_Elapsed_Time, Air_Time, Flights, Distance,

In [16]:
# No dups!

Now I will address missing and/or null values.

In [18]:
# Set the display option to show all rows
pd.set_option('display.max_rows', None)

In [19]:
# Find the count of null values in each column
null_counts = df.isnull().sum()

In [20]:
# Display the count of null values
print(null_counts)

Year                                                       0
Quarter                                                    0
Month                                                      0
Day_of_Month                                               0
Day_of_Week                                                0
Flight_Date                                                0
Marketing_Airline_Network                                  0
DOT_ID_Marketing_Airline                                   0
IATA_Code_Marketing_Airline                                0
Flight_Number_Marketing_Airline                            0
IATA_Code_Originally_Scheduled_Code_Share_Airline    5021838
Operating_Airline                                          0
DOT_ID_Operating_Airline                                   0
IATA_Code_Operating_Airline                                0
Tail_Number                                           168453
Flight_Number_Operating_Airline                            0
Origin_Airport_ID       

In [21]:
# The IATA_Code_Originally_Scheduled_Code_Share column is nearly completely null
# Given that the Bureau of Transportation Statistics (BTS.gov) stated that this column shouldn't be 
# used for analysis, and it is completely null for 2019, I will be dropping it and keeping the Unique_Code
# column as suggested by BTS.gov
# Here is the language from BTS.gov:
# ""Code assigned by IATA and commonly used to identify a carrier.""
#""As the same code may have been assigned to different carriers over time, the code is not always unique.""
#""For analysis, use the Unique Carrier Code.""

# Drop the 'IATA_Code_Originally_Scheduled_Code_Share_Airline' column
df = df.drop(columns=['IATA_Code_Originally_Scheduled_Code_Share_Airline'])

In [22]:
df.shape

(5022397, 64)

There are a significant amount of null values for Dep_Delay, Taxi_Out, Taxi_In, Arr_Time.
I believe these are null values because the flights were cancelled.
I will check value counts for the Cancelled column. Value counts for '1' should be close if not the same to the values missing.

In [23]:
df['Cancelled'].value_counts()

Cancelled
0.0    4721342
1.0     301055
Name: count, dtype: int64

301,055 flights were cancelled in 2020.

This is fairly close to the missing values in:

Dep_Time                                              299620
Dep_Delay                                             299683
Dep_Delay_Minutes                                     299683
Dep_Del_15                                            299683
Departure_Delay_Groups                                29
Taxi_Out                                              300426
Wheels_Off                                            300426
Wheels_On                                             302000
Taxi_In                                               302
Arr_Time                                              301999
Arr_Delay                                             309467
Arr_Delay_Minutes                                     309467
Arr_Del_15                                            309467
Arrival_Delay_Groups                                  30

The missing values in:

Cancellation_Code                                    4721342 -- are accounted for because it is the same as the total Flights (5022397) minus the cancelled flights (301,055) - which is 4721342. Therefore 4721342 flights have no cancellation code because they were not cancelled.

After careful consideration, I believe it is best to create a cancelled flights flag in order to prevent confusion in future analysis for delays v. cancellation. 94670009683

In [24]:
# Identify rows where flights were cancelled

cancelled_flights = df['Cancelled'] == 1

In [25]:
# Create a column to indicate if the data is missing due to cancellation

df['Missing_Due_To_Cancellation'] = cancelled_flights

Now I will address columns containing mixed datatypes.

In [26]:
df.dtypes

Year                                 int64
Quarter                              int64
Month                                int64
Day_of_Month                         int64
Day_of_Week                          int64
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int64
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int64
Operating_Airline                   object
DOT_ID_Operating_Airline             int64
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int64
Origin_Airport_ID                    int64
Origin_Airport_Seq_ID                int64
Origin_City_Market_ID                int64
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                    int64
Origin_State_Name                   object
Dest_Airpor

In [27]:
# I will change many of these datatypes to reduce the amount of memory needed by reducing int64 to int8, 16, or 32 where possible.
# Additionally, I will change the Flight_Date column to a date datatype, instead of an object.
# Lets change Flight_Date first

# Ensure 'Flight_Date' column is in datetime format
df['Flight_Date'] = pd.to_datetime(df['Flight_Date'], errors='coerce')

In [28]:
# Extract only the date part (since we don't have time in the same column)
df['Flight_Date'] = df['Flight_Date'].dt.date

In [29]:
# Verify the result
print(df.head())
print(df.dtypes)

   Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2020        1      1            17            5  2020-01-17   
1  2020        1      1            18            6  2020-01-18   
2  2020        1      1            19            7  2020-01-19   
3  2020        1      1            20            1  2020-01-20   
4  2020        1      1            21            2  2020-01-21   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        B6                     20409   
1                        B6                     20409   
2                        B6                     20409   
3                        B6                     20409   
4                        B6                     20409   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          B6                              942   
1                          B6                              942   
2                          B6                              942

In [30]:
# Change datatypes to maximize available memory

df = df.astype({
    'Year': 'int16',
    'Quarter': 'int16',
    'Month': 'int16',
    'Day_of_Month': 'int16',
    'Day_of_Week': 'int16',
    'DOT_ID_Marketing_Airline': 'int16',
    'Flight_Number_Marketing_Airline': 'int16',
    'DOT_ID_Operating_Airline': 'int16',
    'Flight_Number_Operating_Airline': 'int16',
    'Origin_Airport_ID': 'int16',
    'Origin_Airport_Seq_ID': 'int16',
    'Origin_City_Market_ID': 'int16',
    'Origin_State_Fips': 'int8',
    'Dest_Airport_ID': 'int16',
    'Dest_Airport_Seq_ID': 'int16',
    'Dest_City_Market_ID': 'int16',
    'Dest_State_Fips': 'int16',
    'CRS_Dep_Time': 'int16'
})

In [31]:
df.dtypes

Year                                 int16
Quarter                              int16
Month                                int16
Day_of_Month                         int16
Day_of_Week                          int16
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int16
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int16
Operating_Airline                   object
DOT_ID_Operating_Airline             int16
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int16
Origin_Airport_ID                    int16
Origin_Airport_Seq_ID                int16
Origin_City_Market_ID                int16
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                     int8
Origin_State_Name                   object
Dest_Airpor

In [32]:
df.shape

(5022397, 65)

In [33]:
df.head()

Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2020        1      1            17            5  2020-01-17   
1  2020        1      1            18            6  2020-01-18   
2  2020        1      1            19            7  2020-01-19   
3  2020        1      1            20            1  2020-01-20   
4  2020        1      1            21            2  2020-01-21   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        B6                     20409   
1                        B6                     20409   
2                        B6                     20409   
3                        B6                     20409   
4                        B6                     20409   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          B6                              942   
1                          B6                              942   
2                          B6                              942   
3                          B6                              942   
4                          B6                              942   

  Operating_Airline   DOT_ID_Operating_Airline IATA_Code_Operating_Airline  \
0                 B6                     20409                          B6   
1                 B6                     20409                          B6   
2                 B6                     20409                          B6   
3                 B6                     20409                          B6   
4                 B6                     20409                          B6   

  Tail_Number  Flight_Number_Operating_Airline  Origin_Airport_ID  \
0      N661JB                              942              14027   
1      N584JB                              942              14027   
2      N662JB                              942              14027   
3      N784JB                              942              14027   
4      N784JB                              942              14027   

   Origin_Airport_Seq_ID  Origin_City_Market_ID Origin  \
0                  26446                 -31509    PBI   
1                  26446                 -31509    PBI   
2                  26446                 -31509    PBI   
3                  26446                 -31509    PBI   
4                  26446                 -31509    PBI   

                 Origin_City_Name Origin_State  Origin_State_Fips  \
0  West Palm Beach/Palm Beach, FL           FL                 12   
1  West Palm Beach/Palm Beach, FL           FL                 12   
2  West Palm Beach/Palm Beach, FL           FL                 12   
3  West Palm Beach/Palm Beach, FL           FL                 12   
4  West Palm Beach/Palm Beach, FL           FL                 12   

  Origin_State_Name  Dest_Airport_ID  Dest_Airport_Seq_ID  \
0           Florida            10529                 4330   
1           Florida            10529                 4330   
2           Florida            10529                 4330   
3           Florida            10529                 4330   
4           Florida            10529                 4330   

   Dest_City_Market_ID Dest Dest_City_Name Dest_State  Dest_State_Fips  \
0                30529  BDL   Hartford, CT         CT                9   
1                30529  BDL   Hartford, CT         CT                9   
2                30529  BDL   Hartford, CT         CT                9   
3                30529  BDL   Hartford, CT         CT                9   
4                30529  BDL   Hartford, CT         CT                9   

  Dest_State_Name  CRS_Dep_Time  Dep_Time  Dep_Delay  Dep_Delay_Minutes  \
0     Connecticut          1940    1926.0      -14.0                0.0   
1     Connecticut          1940    1926.0      -14.0                0.0   
2     Connecticut          1940    1947.0        7.0                7.0   
3     Connecticut          1940    1933.0       -7.0                0.0   
4     Connecticut          1940

In [34]:
# Save Clean 2020 Data

df.to_csv(os.path.join(path, 'clean_flights2020.csv'))